File: isu-sevice-maven/src/main/java/com/example/isusevicemaven/StudyGroupApi.java
package com.example.isusevicemaven;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import entity.StudyGroup;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpMethod;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;

import java.net.URI;

@Service
public class StudyGroupApi {

    private final RestTemplate restTemplate;
    private final String host;
    private final ObjectMapper objectMapper;

    public StudyGroupApi(RestTemplate restTemplate, ObjectMapper objectMapper) {
        this.restTemplate = restTemplate;
        this.host = "http://study-group-service";
        this.objectMapper = objectMapper;
    }

    public StudyGroup studyGroupById(int id) throws JsonProcessingException {
        var exchange = restTemplate.exchange(
                URI.create(host + "/groups/" + id),
                HttpMethod.GET,
                null,
                String.class
        );
        return objectMapper.readValue(exchange.getBody(), StudyGroup.class);
    }

    public StudyGroup putStudyGroup(StudyGroup studyGroup) throws JsonProcessingException {
            var exchange = restTemplate.exchange(
                    URI.create(host + "/groups/" + studyGroup.getId()),
                    HttpMethod.PUT,
                    new HttpEntity<>(studyGroup),
                    String.class
            );
            return objectMapper.readValue(exchange.getBody(), StudyGroup.class);
    }
}

File: isu-service/src/main/kotlin/com/example/isuservice/service/StudyGroupApi.kt
package com.example.isuservice.service

import com.fasterxml.jackson.databind.ObjectMapper
import entity.StudyGroup
import org.springframework.http.HttpEntity
import org.springframework.http.HttpMethod
import org.springframework.stereotype.Service
import org.springframework.web.client.HttpClientErrorException
import org.springframework.web.client.RestTemplate
import java.net.URI

@Service
class StudyGroupApi(
    val restTemplate: RestTemplate,
    val host: String = "http://study-group-service",
    val objectMapper: ObjectMapper
) {
    fun studyGroupById(id: Int): Result =
        try {
            val exchange = restTemplate.exchange(
                URI("$host/groups/$id"),
                HttpMethod.GET,
                null,
                String::class.java
            )
            val readValue = objectMapper.readValue(exchange.body, StudyGroup::class.java)
            Result.Success(readValue)
        } catch (e: HttpClientErrorException) {
            Result.Error(e.statusCode, e.message)
        }


    fun putStudyGroup(studyGroup: StudyGroup): Result =
        try {
            val exchange = restTemplate.exchange(
                URI("$host/groups/${studyGroup.id}"),
                HttpMethod.PUT,
                HttpEntity(studyGroup),
                String::class.java
            )
            val readValue = objectMapper.readValue(exchange.body, StudyGroup::class.java)
            Result.Success(readValue)
        } catch (e: HttpClientErrorException) {
            Result.Error(e.statusCode, e.message)
        }
}